node/src/pruner.rs

```diff
@@ -106,12 +117,25 @@ where
     S: Store + 'static,
     B: Blockstore + 'static,
 {
+    daser: Arc<Daser>,
     cancellation_token: CancellationToken,
     event_pub: EventPublisher,
     store: Arc<S>,
     blockstore: Arc<B>,
-    pruning_interval: Duration,
+    block_time: Duration,
     pruning_window: Duration,
+    sampling_window: Duration,
+    prev_num_of_prunable_blocks: u64,
+    cache: Cache,
+}
+
+struct Cache {
+    /// When Cache was updated.
+    updated_at: Option<Instant>,
+    /// The first known height that is after the pruning window.
+    after_pruning_window: Option<u64>,
+    /// The first known height that is after the sampling window.
+    after_sampling_window: Option<u64>,
```

Comment:
If I understand correctly, maybe using `stored` instead of `known` would be more accurate here?

Reply:
not really. let say we `after_pruning_window` is height 150 and it gets pruned, the next height that is after pruning window will be 149, but because if is smaller than 150 we do not update the value. so `after_pruning_window` holds a height that is known but not stored.